File: frontend/src-tauri/src/recall/indexer.rs
//! Build and refresh the recall index. Indexing is idempotent (a re-run over unchanged
//! transcript text with the same embedder is a no-op) and best-effort about embeddings:
//! if the local embedder is unavailable, the meeting is indexed lexical-only and will be
//! upgraded to embeddings on a later run once the model is present.

use std::sync::atomic::{AtomicBool, Ordering};

use chrono::Utc;
use sqlx::SqlitePool;
use uuid::Uuid;

use crate::database::repositories::{
    meeting::MeetingsRepository,
    recall_index::{RecallChunkInput, RecallIndexRepository},
};
use crate::recall::{chunker, embedding};

/// Guards against overlapping full backfills (startup + first-query auto-trigger + the
/// explicit reindex command can all race). Per-meeting `index_meeting` is unguarded and
/// cheap.
static REINDEX_RUNNING: AtomicBool = AtomicBool::new(false);

fn fnv1a_hex(text: &str) -> String {
    let mut hash: u64 = 0xcbf29ce484222325;
    for byte in text.as_bytes() {
        hash ^= *byte as u64;
        hash = hash.wrapping_mul(0x0000_0100_0000_01b3);
    }
    format!("{hash:016x}")
}

/// Index one meeting. Logs its own errors and never panics — safe to `spawn` fire-and-forget.
pub async fn index_meeting(pool: &SqlitePool, meeting_id: &str) {
    if let Err(error) = index_meeting_inner(pool, meeting_id).await {
        log::warn!("recall: failed to index meeting {meeting_id}: {error}");
    }
}

async fn index_meeting_inner(pool: &SqlitePool, meeting_id: &str) -> Result<(), String> {
    let (transcripts, _total) =
        MeetingsRepository::get_meeting_transcripts_paginated(pool, meeting_id, i64::MAX, 0)
            .await
            .map_err(|e| e.to_string())?;

    let joined: String = transcripts
        .iter()
        .map(|t| t.transcript.trim())
        .filter(|t| !t.is_empty())
        .collect::<Vec<_>>()
        .join("\n");
    if joined.trim().is_empty() {
        // No transcript text (e.g. summary-only meeting) — clear any stale index.
        let _ = RecallIndexRepository::delete_meeting(pool, meeting_id).await;
        return Ok(());
    }

    let content_hash = fnv1a_hex(&joined);
    let want_model = embedding::current_model_tag(pool).await;

    // Idempotency: skip only when unchanged AND already fully embedded with this model.
    // A prior lexical-only index (embedded_count < chunk_count) is intentionally re-run so
    // embeddings fill in once the model becomes available.
    if let Ok(Some(state)) = RecallIndexRepository::get_index_state(pool, meeting_id).await {
        if state.content_hash == content_hash
            && state.chunk_count > 0
            && state.embedded_count == state.chunk_count
            && state.embedding_model.as_deref() == Some(want_model.as_str())
        {
            return Ok(());
        }
    }

    let drafts = chunker::chunk_transcripts(&transcripts);
    if drafts.is_empty() {
        let _ = RecallIndexRepository::delete_meeting(pool, meeting_id).await;
        return Ok(());
    }

    // Embed all chunks in one batch with the configured backend. Best-effort: any failure
    // (embedder unavailable, or a mismatched count) falls back to lexical-only for the whole
    // meeting — never a partial or fabricated set of vectors.
    let texts: Vec<String> = drafts.iter().map(|d| d.text.clone()).collect();
    let embeddings: Option<Vec<Vec<f32>>> = match embedding::embed_documents(pool, &texts).await {
        Ok(vectors) if vectors.len() == texts.len() => Some(vectors),
        Ok(_) => {
            log::warn!(
                "recall: embedder returned a mismatched count; indexing {meeting_id} lexical-only"
            );
            None
        }
        Err(error) => {
            log::warn!(
                "recall: embedder unavailable ({error}); indexing meeting {meeting_id} lexical-only"
            );
            None
        }
    };

    let mut inputs = Vec::with_capacity(drafts.len());
    for (index, draft) in drafts.iter().enumerate() {
        let (embedding_bytes, dim) = match &embeddings {
            Some(vectors) => (
                Some(embedding::pack_f32(&vectors[index])),
                Some(vectors[index].len() as i64),
            ),
            None => (None, None),
        };
        inputs.push(RecallChunkInput {
            id: Uuid::new_v4().to_string(),
            chunk_index: draft.chunk_index,
            chunk_text: draft.text.clone(),
            start_time: draft.start_time,
            end_time: draft.end_time,
            timestamp_label: draft.timestamp_label.clone(),
            embedding_model: dim.map(|_| want_model.clone()),
            dim,
            embedding: embedding_bytes,
            token_estimate: Some(draft.token_estimate),
        });
    }

    let model_used = embeddings.as_ref().map(|_| want_model.as_str());
    RecallIndexRepository::replace_meeting_chunks(
        pool,
        meeting_id,
        &inputs,
        &content_hash,
        model_used,
        &Utc::now().to_rfc3339(),
    )
    .await
    .map_err(|e| e.to_string())?;

    Ok(())
}

/// True if a backfill was started by this call; false if one is already running.
pub fn try_begin_reindex() -> bool {
    !REINDEX_RUNNING.swap(true, Ordering::SeqCst)
}

pub fn end_reindex() {
    REINDEX_RUNNING.store(false, Ordering::SeqCst);
}

/// Read-only check of whether a backfill is in progress (does not acquire the guard).
pub fn is_reindex_running() -> bool {
    REINDEX_RUNNING.load(Ordering::SeqCst)
}

/// Backfill every meeting that is missing or stale. Self-guarded against overlap: returns
/// `Ok(0)` immediately if a backfill is already running. `force` re-indexes even unchanged
/// meetings (e.g. to embed a vault that was previously indexed lexical-only).
pub async fn reindex_all(pool: &SqlitePool, force: bool) -> Result<usize, String> {
    if !try_begin_reindex() {
        return Ok(0);
    }
    let result = reindex_all_inner(pool, force).await;
    end_reindex();
    result
}

async fn reindex_all_inner(pool: &SqlitePool, force: bool) -> Result<usize, String> {
    let meetings = MeetingsRepository::get_meetings(pool)
        .await
        .map_err(|e| e.to_string())?;
    let mut indexed = 0usize;
    for meeting in meetings {
        if force {
            let _ = RecallIndexRepository::delete_meeting(pool, &meeting.id).await;
        }
        index_meeting(pool, &meeting.id).await;
        indexed += 1;
    }
    Ok(indexed)
}

File: AriKit/Sources/AriKit/Recall/Indexer/Indexer.swift
//
//  Indexer.swift — build and refresh the recall index (plan §5 Slice 5, ← indexer.rs).
//
//  Indexing is idempotent (a re-run over unchanged transcript text with the same embedder is a
//  no-op) and best-effort about embeddings: if the local embedder is unavailable, the meeting is
//  indexed lexical-only and will be upgraded to embeddings on a later run once the model is
//  present. No Settings layer exists yet (plan §9 decision 1), so "the current model tag" is
//  simply the injected embedder's `modelTag` — there is nothing to invent here.
//
import Foundation
import os

/// Builds/refreshes the recall index for one or all meetings. `Sendable` value type over
/// injected repository handles + an embedder + a shared `ReindexCoordinator` — safe to call from
/// any isolation domain (mirrors `HybridSearch`'s shape).
public struct Indexer: Sendable {
    private static let logger = Logger(subsystem: "com.arivo.ari.AriKit", category: "recall.indexer")

    private let recallIndex: RecallIndexRepository
    private let transcripts: TranscriptRepository
    private let meetings: MeetingRepository
    private let summaries: SummaryRepository
    private let embedder: any RecallEmbedder
    private let coordinator: ReindexCoordinator

    public init(
        recallIndex: RecallIndexRepository,
        transcripts: TranscriptRepository,
        meetings: MeetingRepository,
        summaries: SummaryRepository,
        embedder: any RecallEmbedder,
        coordinator: ReindexCoordinator
    ) {
        self.recallIndex = recallIndex
        self.transcripts = transcripts
        self.meetings = meetings
        self.summaries = summaries
        self.embedder = embedder
        self.coordinator = coordinator
    }

    /// Index one meeting. Logs its own errors and NEVER throws — safe to fire-and-forget from a
    /// detached `Task` after a save (← `index_meeting`, indexer.rs:33).
    public func indexMeeting(_ meetingId: MeetingID) async {
        do {
            try await indexMeetingInner(meetingId)
        } catch {
            Self.logger.warning(
                "recall: failed to index meeting \(meetingId.rawValue, privacy: .public): \(String(describing: error), privacy: .public)"
            )
        }
    }

    /// ← `index_meeting_inner` (indexer.rs:39-135), extended by plan
    /// `ask-meetings-tools-and-cards.md` §3.2 (Bug B fix): the summary body is now a SECOND chunk
    /// stream, tagged `sourceKind == .summary` so a summary-only fact becomes searchable even
    /// though it never appears verbatim in the raw transcript. The content hash covers BOTH texts,
    /// so a newly-generated/edited summary re-triggers indexing even when the transcript itself is
    /// unchanged.
    private func indexMeetingInner(_ meetingId: MeetingID) async throws {
        let transcriptRows = try await transcripts.forMeeting(meetingId)
        let summary = try await summaries.forMeeting(meetingId)

        let joinedTranscript = transcriptRows
            .map { $0.transcript.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .joined(separator: "\n")
        let summaryText = summary?.bodyMarkdown.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        if joinedTranscript.isEmpty, summaryText.isEmpty {
            // Nothing to index at all — clear any stale index.
            try? await recallIndex.deleteMeeting(meetingId)
            return
        }

        let contentHash = Self.fnv1aHex(joinedTranscript + "\n---\n" + summaryText)
        let wantModel = embedder.modelTag

        // Idempotency: skip only when unchanged AND already fully embedded with this model. A
        // prior lexical-only index (embeddedCount < chunkCount) is intentionally re-run so
        // embeddings fill in once the model becomes available.
        if let state = try? await recallIndex.indexState(meetingId: meetingId),
           state.contentHash == contentHash,
           state.chunkCount > 0,
           state.embeddedCount == state.chunkCount,
           state.embeddingModel == wantModel {
            return
        }

        let transcriptDrafts = Recall.chunkTranscripts(transcriptRows)
        let summaryDrafts = summaryText.isEmpty ? [] : Recall.chunkSummary(summaryText)
        if transcriptDrafts.isEmpty, summaryDrafts.isEmpty {
            try? await recallIndex.deleteMeeting(meetingId)
            return
        }

        // Tag each draft with the stream it came from BEFORE embedding, so the two arrays can be
        // combined into one `embedder.embed(texts)` call (embedding-model consistency, fewer
        // round trips) and split back apart afterward.
        let taggedDrafts: [(draft: ChunkDraft, sourceKind: RecallChunkSourceKind)] =
            transcriptDrafts.map { ($0, .transcript) } + summaryDrafts.map { ($0, .summary) }

        // Embed all chunks (transcript + summary together) in one batch with the injected backend.
        // Best-effort: any failure (embedder unavailable, or a mismatched count) falls back to
        // lexical-only for the whole meeting — never a partial or fabricated set of vectors.
        let texts = taggedDrafts.map(\.draft.text)
        let embeddings: [[Float]]?
        do {
            let vectors = try await embedder.embed(texts)
            if vectors.count == texts.count {
                embeddings = vectors
            } else {
                Self.logger.warning(
                    "recall: embedder returned a mismatched count; indexing \(meetingId.rawValue, privacy: .public) lexical-only"
                )
                embeddings = nil
            }
        } catch {
            Self.logger.warning(
                "recall: embedder unavailable (\(String(describing: error), privacy: .public)); indexing meeting \(meetingId.rawValue, privacy: .public) lexical-only"
            )
            embeddings = nil
        }

        var inputs: [RecallChunkInput] = []
        inputs.reserveCapacity(taggedDrafts.count)
        for (index, tagged) in taggedDrafts.enumerated() {
            let draft = tagged.draft
            let embeddingBytes: Data?
            let dim: Int?
            if let vectors = embeddings {
                embeddingBytes = Recall.packF32(vectors[index])
                dim = vectors[index].count
            } else {
                embeddingBytes = nil
                dim = nil
            }
            inputs.append(RecallChunkInput(
                id: RecallChunkID(UUID().uuidString),
                chunkIndex: draft.chunkIndex,
                chunkText: draft.text,
                startTime: draft.startTime,
                endTime: draft.endTime,
                timestampLabel: draft.timestampLabel,
                embedding: embeddingBytes,
                embeddingModel: dim != nil ? wantModel : nil,
                dim: dim,
                tokenEstimate: draft.tokenEstimate,
                sourceKind: tagged.sourceKind
            ))
        }

        let modelUsed = embeddings != nil ? wantModel : nil
        try await recallIndex.replaceMeetingChunks(
            meetingId: meetingId,
            chunks: inputs,
            contentHash: contentHash,
            embeddingModel: modelUsed,
            now: RFC3339.string(from: Date())
        )
    }

    /// Backfill every meeting that is missing or stale. Self-guarded against overlap via the
    /// shared `ReindexCoordinator`: returns `0` immediately if a backfill is already running.
    /// `force` re-indexes even unchanged meetings (e.g. to embed a vault that was previously
    /// indexed lexical-only) (← `reindex_all`, indexer.rs:154-161).
    public func reindexAll(force: Bool) async throws -> Int {
        guard await coordinator.tryBegin() else {
            return 0
        }
        do {
            let indexed = try await reindexAllInner(force: force)
            await coordinator.end()
            return indexed
        } catch {
            await coordinator.end()
            throw error
        }
    }

    /// ← `reindex_all_inner` (indexer.rs:163-176).
    private func reindexAllInner(force: Bool) async throws -> Int {
        let allMeetings = try await meetings.all()
        var indexed = 0
        for meeting in allMeetings {
            if force {
                try? await recallIndex.deleteMeeting(meeting.id)
            }
            await indexMeeting(meeting.id)
            indexed += 1
        }
        return indexed
    }

    /// FNV-1a (64-bit) hex digest of `text`'s UTF-8 bytes — the exact algorithm/seed from
    /// `indexer.rs:23-30`, so hashes computed by either side match byte-for-byte.
    static func fnv1aHex(_ text: String) -> String {
        var hash: UInt64 = 0xCBF2_9CE4_8422_2325
        for byte in text.utf8 {
            hash ^= UInt64(byte)
            hash = hash &* 0x0000_0100_0000_01B3
        }
        var hex = String(hash, radix: 16)
        if hex.count < 16 {
            hex = String(repeating: "0", count: 16 - hex.count) + hex
        }
        return hex
    }
}
